File: app/src/main/java/ch/protonmail/android/views/PMWebViewClient.java
/*
 * Copyright (c) 2020 Proton Technologies AG
 * 
 * This file is part of ProtonMail.
 * 
 * ProtonMail is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * ProtonMail is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with ProtonMail. If not, see https://www.gnu.org/licenses/.
 */
package ch.protonmail.android.views;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.MailTo;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.text.Html;
import android.text.Spanned;
import android.text.TextUtils;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.annotation.NonNull;

import java.io.ByteArrayInputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Arrays;
import java.util.List;

import ch.protonmail.android.R;
import ch.protonmail.android.activities.composeMessage.ComposeMessageActivity;
import ch.protonmail.android.api.models.User;
import ch.protonmail.android.core.Constants;
import ch.protonmail.android.core.UserManager;
import ch.protonmail.android.utils.AppUtil;
import ch.protonmail.android.utils.MessageUtils;
import ch.protonmail.android.utils.extensions.TextExtensions;
import ch.protonmail.android.utils.ui.dialogs.DialogUtils;

import static ch.protonmail.android.activities.composeMessage.ComposeMessageActivity.EXTRA_MAIL_TO;
import static ch.protonmail.android.activities.composeMessage.ComposeMessageActivity.EXTRA_MESSAGE_BODY;
import static ch.protonmail.android.activities.composeMessage.ComposeMessageActivity.EXTRA_MESSAGE_TITLE;

public class PMWebViewClient extends WebViewClient {

    private final UserManager mUserManager;
    private final Activity mActivity;
    private boolean mLoadRemote;
    private int mBlockedImages;

    private final List<String> hyperlinkConfirmationWhitelistedHosts = Arrays.asList(
            "protonmail.com",
            "protonmail.ch",
            "protonvpn.com",
            "protonstatus.com",
            "gdpr.eu",
            "protonvpn.net",
            "pm.me",
            "mail.protonmail.com",
            "account.protonvpn.com",
            "protonirockerxow.onion"
    );

    public PMWebViewClient(
            @NonNull UserManager userManager,
            @NonNull Activity activity,
            boolean loadRemote
    ) {
        mUserManager = userManager;
        mActivity = activity;
        mLoadRemote = loadRemote;
        mBlockedImages = 0;
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url1) {
        String url = url1.replaceFirst(Constants.DUMMY_URL_PREFIX, "");

        if (url.startsWith("mailto:")) {
            composeMessageWithMailToData(url, view.getContext().getApplicationContext());
            return true;
        }

        if (url.startsWith("tel:")) {
            Intent intent = new Intent(Intent.ACTION_DIAL);
            intent.setData(Uri.parse(url));
            if (intent.resolveActivity(mActivity.getPackageManager()) != null) {
                mActivity.startActivity(intent);
            } else {
                TextExtensions.showToast(mActivity, R.string.no_application_found);
            }
        } else {
            if (!TextUtils.isEmpty(url)) {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(url));
                if (intent.resolveActivity(mActivity.getPackageManager()) != null) {
                    if (!showHyperlinkConfirmation(url)) {
                        mActivity.startActivity(intent);
                    }
                } else {
                    TextExtensions.showToast(mActivity, R.string.no_application_found_or_link_invalid);
                }
            }
        }
        return true;
    }

    private void composeMessageWithMailToData(String url, Context context) {
        Intent intent = AppUtil.decorInAppIntent(
                new Intent(context, ComposeMessageActivity.class)
        );
        MailTo mt = MailTo.parse(url);

        User user = mUserManager.getUser();
        MessageUtils.INSTANCE.addRecipientsToIntent(
                intent,
                ComposeMessageActivity.EXTRA_TO_RECIPIENTS,
                mt.getTo(),
                Constants.MessageActionType.FROM_URL,
                user.getAddresses()
        );
        MessageUtils.INSTANCE.addRecipientsToIntent(intent,
                ComposeMessageActivity.EXTRA_CC_RECIPIENTS,
                mt.getCc(),
                Constants.MessageActionType.FROM_URL,
                user.getAddresses()
        );
        intent.putExtra(EXTRA_MAIL_TO, true);
        intent.putExtra(EXTRA_MESSAGE_TITLE, mt.getSubject());
        intent.putExtra(EXTRA_MESSAGE_BODY, mt.getBody());
        mActivity.startActivity(intent);
    }

    /**
     *
     * @return true if confirmation was shown
     */
    private boolean showHyperlinkConfirmation(@NonNull String url) {

        if (!Constants.FeatureFlags.HYPERLINK_CONFIRMATION) return false;
        try {
            URL parsedUrl = new URL(url);

            // only handle http/s protocols
            if (!"http".equals(parsedUrl.getProtocol().toLowerCase()) && !"https".equals(parsedUrl.getProtocol().toLowerCase())) return false;

            // whitelist domains
            for (String host: hyperlinkConfirmationWhitelistedHosts) {
                if (parsedUrl.getHost().endsWith(host)) return false;
            }
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }

        if (PreferenceManager.getDefaultSharedPreferences(mActivity).getBoolean(Constants.Prefs.PREF_HYPERLINK_CONFIRM, true)) {

            String hyperlink = url; // truncate if too long
            if (hyperlink.length() > 100) {
                hyperlink = hyperlink.substring(0, 60) + "..." + hyperlink.substring(hyperlink.length() - 40);
            }
            Spanned message = Html.fromHtml(String.format(mActivity.getString(R.string.hyperlink_confirmation_dialog_text_html), hyperlink));

            DialogUtils.Companion.showInfoDialogWithTwoButtonsAndCheckbox(mActivity, "", message, mActivity.getString(R.string.cancel), mActivity.getString(R.string.cont), mActivity.getString(R.string.dont_ask_again), unit -> {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(url));
                mActivity.startActivity(intent);
                return null;
            }, isChecked -> {
                PreferenceManager.getDefaultSharedPreferences(mActivity).edit().putBoolean(Constants.Prefs.PREF_HYPERLINK_CONFIRM, !isChecked).apply();
                return null;
            }, true);

            return true;
        } else {
            return false;
        }
    }

    protected int amountOfRemoteResourcesBlocked() {
        return mBlockedImages;
    }

    public void blockRemoteResources(boolean block) {
        mBlockedImages = 0;
        mLoadRemote = !block;
    }

    public void loadRemoteResources() {
        blockRemoteResources(false);
    }

    @Override
    public WebResourceResponse shouldInterceptRequest(WebView view, String url) {
        if (mLoadRemote) {
            return super.shouldInterceptRequest(view, url);
        }
        Uri uri = Uri.parse(url);
        if (uri.getScheme().equalsIgnoreCase("cid") || uri.getScheme().equalsIgnoreCase("data")) {
            return super.shouldInterceptRequest(view, url);
        }
        if (url.toLowerCase().contains("/favicon.ico")) {
            return super.shouldInterceptRequest(view, url);
        }
        mBlockedImages++;
        return new WebResourceResponse("text/plain" , "utf-8", new ByteArrayInputStream(new byte[0]));
    }
}

File: app/src/main/java/ch/protonmail/android/views/PmWebViewClient.kt
/*
 * Copyright (c) 2022 Proton AG
 *
 * This file is part of Proton Mail.
 *
 * Proton Mail is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Proton Mail is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Proton Mail. If not, see https://www.gnu.org/licenses/.
 */
package ch.protonmail.android.views

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.MailTo
import android.net.Uri
import android.webkit.WebResourceResponse
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.core.text.HtmlCompat
import ch.protonmail.android.R
import ch.protonmail.android.activities.composeMessage.ComposeMessageActivity
import ch.protonmail.android.core.Constants
import ch.protonmail.android.core.Constants.DUMMY_URL_PREFIX
import ch.protonmail.android.core.UserManager
import ch.protonmail.android.settings.data.AccountSettingsRepository
import ch.protonmail.android.utils.MessageUtils.addRecipientsToIntent
import ch.protonmail.android.utils.ui.dialogs.DialogUtils.Companion.showInfoDialogWithTwoButtonsAndCheckbox
import ch.protonmail.android.utils.ui.dialogs.DialogUtils.Companion.showTwoButtonInfoDialog
import kotlinx.coroutines.runBlocking
import me.proton.core.presentation.utils.showToast
import me.proton.core.util.kotlin.startsWith
import timber.log.Timber
import java.io.ByteArrayInputStream
import java.net.MalformedURLException
import java.net.URL
import java.util.Locale

open class PmWebViewClient(
    private val userManager: UserManager,
    private val accountSettingsRepository: AccountSettingsRepository,
    private val activity: Activity,
    private var shouldLoadRemoteContent: Boolean
) : WebViewClient() {

    private var blockedImages = 0
    private var isPhishingMessage = false
    private val hyperlinkConfirmationWhitelistedHosts = listOf(
        "protonmail.com",
        "protonmail.ch",
        "protonvpn.com",
        "protonstatus.com",
        "gdpr.eu",
        "protonvpn.net",
        "pm.me",
        "mail.protonmail.com",
        "account.protonvpn.com",
        "protonirockerxow.onion",
        "proton.me"
    )

    @Deprecated("Deprecated in Java")
    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        val fixedUrl = url.replaceFirst(DUMMY_URL_PREFIX.toRegex(), "")
        if (fixedUrl startsWith "mailto:") {
            composeMessageWithMailToData(fixedUrl, view.context.applicationContext)
            return true
        }
        if (fixedUrl startsWith "tel:") {
            val intent = Intent(Intent.ACTION_DIAL)
            intent.data = Uri.parse(fixedUrl)
            if (intent.resolveActivity(activity.packageManager) != null) {
                activity.startActivity(intent)
            } else {
                activity.showToast(R.string.no_application_found)
            }
        } else {
            if (fixedUrl.isNotBlank()) {
                val intent = Intent(Intent.ACTION_VIEW)
                intent.data = Uri.parse(fixedUrl)
                if (showHyperlinkConfirmation(fixedUrl).not()) {
                    try {
                        activity.startActivity(intent)
                    } catch (notFoundException: ActivityNotFoundException) {
                        Timber.i(notFoundException, "Unable to open link")
                        activity.showToast(R.string.no_application_found)
                    }
                }
            }
        }
        return true
    }

    fun setPhishingCheck(isPhishingMessage: Boolean) {
        this.isPhishingMessage = isPhishingMessage
    }

    private fun composeMessageWithMailToData(url: String, context: Context) {
        val intent = Intent(context, ComposeMessageActivity::class.java)

        val mailTo = MailTo.parse(url)
        val user = userManager.currentUser
            ?: return
        addRecipientsToIntent(
            intent = intent,
            extraName = ComposeMessageActivity.EXTRA_TO_RECIPIENTS,
            recipientList = mailTo.to,
            messageAction = Constants.MessageActionType.FROM_URL,
            userAddresses = user.addresses
        )
        addRecipientsToIntent(
            intent = intent,
            extraName = ComposeMessageActivity.EXTRA_CC_RECIPIENTS,
            recipientList = mailTo.cc,
            messageAction = Constants.MessageActionType.FROM_URL,
            userAddresses = user.addresses
        )
        intent.putExtra(ComposeMessageActivity.EXTRA_MAIL_TO, true)
            .putExtra(ComposeMessageActivity.EXTRA_MESSAGE_TITLE, mailTo.subject)
            .putExtra(ComposeMessageActivity.EXTRA_MESSAGE_BODY, mailTo.body)
        activity.startActivity(intent)
    }

    /**
     * @return true if confirmation was shown
     */
    private fun showHyperlinkConfirmation(url: String): Boolean {
        try {
            val parsedUrl = URL(url)

            // only handle http/s protocols
            val protocol = parsedUrl.protocol.lowercase(Locale.getDefault())
            if (protocol != "http" && protocol != "https") return false

            // whitelist domains
            for (host in hyperlinkConfirmationWhitelistedHosts) {
                if (parsedUrl.host.endsWith(host)) return false
            }
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }
        val doesRequireHyperlinkConfirmation = runBlocking {
            accountSettingsRepository
                .getShouldShowLinkConfirmationSetting(userManager.requireCurrentUserId())
        }

        return when {
            isPhishingMessage -> {
                showPhishingHyperlinkConfirmation(url)
                true
            }
            doesRequireHyperlinkConfirmation -> {
                showRegularHyperlinkConfirmation(url)
                true
            }
            else -> {
                false
            }
        }
    }

    private fun showRegularHyperlinkConfirmation(url: String) {
        val message = HtmlCompat.fromHtml(
            activity.getString(R.string.hyperlink_confirmation_dialog_text_html, ellipsesUrlIfTooLong(url)),
            HtmlCompat.FROM_HTML_MODE_LEGACY
        )
        showInfoDialogWithTwoButtonsAndCheckbox(
            context = activity,
            title = "",
            message = message,
            negativeBtnText = activity.getString(R.string.cancel),
            positiveBtnText = activity.getString(
                R.string.cont
            ),
            checkBoxText = activity.getString(R.string.dont_ask_again),
            okListener = {
                val intent = Intent(Intent.ACTION_VIEW)
                intent.data = Uri.parse(url)
                try {
                    activity.startActivity(intent)
                } catch (notFoundException: ActivityNotFoundException) {
                    Timber.i(notFoundException, "Unable to open link")
                    activity.showToast(R.string.no_application_found)
                    it.dismiss()
                }
            },
            checkedListener = { isChecked ->
                runBlocking {
                    accountSettingsRepository.saveShouldShowLinkConfirmationSetting(
                        shouldShowHyperlinkConfirmation = isChecked.not(),
                        userId = userManager.requireCurrentUserId()
                    )
                }
            },
            cancelable = true
        )
    }

    private fun showPhishingHyperlinkConfirmation(url: String) {
        val message = HtmlCompat.fromHtml(
            activity.getString(R.string.details_hyperlink_phishing_dialog_content, ellipsesUrlIfTooLong(url)),
            HtmlCompat.FROM_HTML_MODE_LEGACY
        )
        activity.showTwoButtonInfoDialog(
            titleStringId = R.string.details_hyperlink_phishing_dialog_title,
            message = message,
            positiveStringId = R.string.details_hyperlink_phishing_dialog_confirm_action,
            negativeStringId = R.string.details_hyperlink_phishing_dialog_cancel_action
        ) {
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = Uri.parse(url)
            try {
                activity.startActivity(intent)
            } catch (notFoundException: ActivityNotFoundException) {
                Timber.i(notFoundException, "Unable to open link")
                activity.showToast(R.string.no_application_found)
            }
        }
    }

    private fun ellipsesUrlIfTooLong(url: String) =
        if (url.length > 100) {
            url.substring(0, 60) + "..." + url.substring(url.length - 40)
        } else {
            url
        }

    protected fun amountOfRemoteResourcesBlocked(): Int =
        blockedImages

    fun showRemoteResources(show: Boolean) {
        if (show) blockedImages = 0
        shouldLoadRemoteContent = show
    }

    @Deprecated("Deprecated in Java")
    override fun shouldInterceptRequest(view: WebView, url: String): WebResourceResponse? {
        if (shouldLoadRemoteContent) {
            return super.shouldInterceptRequest(view, url)
        }
        val uri = Uri.parse(url)
        if (uri.scheme.equals("cid", ignoreCase = true) || uri.scheme.equals("data", ignoreCase = true)) {
            return super.shouldInterceptRequest(view, url)
        }
        blockedImages++
        return WebResourceResponse("text/plain", "utf-8", ByteArrayInputStream(ByteArray(0)))
    }
}
